packages-exp/auth-compat-exp/src/user_credential.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as impl from '@firebase/auth-exp';
+import { SignInWithIdpResponse } from '@firebase/auth-exp/src/api/authentication/idp';
+import { UserCredential } from '@firebase/auth-exp/src/model/user';
+import * as compat from '@firebase/auth-types';
+import * as externs from '@firebase/auth-types-exp';
+import '@firebase/installations';
+import { User } from './user';
+
+function credentialFromResponse(
+  userCredential: UserCredential
+): externs.AuthCredential | null {
+  const { providerId, _tokenResponse } = userCredential;
+  if (!_tokenResponse) {
+    return null;
+  }
+  // Handle phone Auth credential responses, as they have a different format
+  // from other backend responses (i.e. no providerId).
+  if ('temporaryProof' in _tokenResponse && 'phoneNumber' in _tokenResponse) {
+    return impl.PhoneAuthProvider.credentialFromResult(userCredential);
+  }
+  // Email and password is not supported as there is no situation where the
+  // server would return the password to the client.
+  if (!providerId || providerId === externs.ProviderId.PASSWORD) {
+    return null;
+  }
+
+  const {
+    oauthIdToken,
+    oauthAccessToken,
+    oauthTokenSecret,
+    pendingToken,
+    nonce
+  } = _tokenResponse as SignInWithIdpResponse;
+  switch (providerId) {
```

Comment:
Why not just have all of these in the switch statements call Provider.credentialFromResult() like you do for phone auth

Reply:
good idea, I guess we didn't have these before and I was just copying the old implementation... we should add one to OAuthProvider for the last bit here